### PR TITLE
Bezier-rs: Add self_intersection to subpath

### PR DESCRIPTION
**Pull request by Androxium**
This PR adds a naive brute-force implementation of the self_intersections function to the Subpath library. This PR also updates the function description of the intersections function to include the minimum_separation variable.

This PR is a rebase of the work done in #915.

### CODE REVIEW

File: libraries/bezier-rs/src/subpath/solvers.rs [hunk 1]
```diff
@@ -10,35 +10,37 @@ impl Subpath {
...
 	pub fn evaluate(&self, t: SubpathTValue) -> DVec2 {
-		let (segment_index, t) = self.t_value_to_parametric(t);
+		let (segment_index, t) = match t {
+			SubpathTValue::GlobalEuclidean(_) => todo!(),
+			SubpathTValue::Parametric { segment_index, t: segment_t } => (segment_index, segment_t),
+			SubpathTValue::GlobalParametric(_) => self.t_value_to_parametric(t),
+			SubpathTValue::Euclidean { segment_index, t } => todo!(),
+			SubpathTValue::EuclideanWithinError { segment_index, t, error } => todo!(),
+			SubpathTValue::GlobalEuclideanWithinError { t, error } => todo!(),
+		};
```
Comment (RobNadal) [inline libraries/bezier-rs/src/subpath/solvers.rs:20]: This whole match statement can be removed

Comment (RobNadal) [inline libraries/bezier-rs/src/subpath/solvers.rs:20]: This whole match statement can be removed

File: libraries/bezier-rs/src/subpath/solvers.rs [hunk 2]
```diff
@@ -52,7 +54,32 @@ impl Subpath {
...
 	}
 
-	/// Returns a normalized unit vector representing the direction of the normal on the subpath based on the parametric `t`-value provided.
+	/// Returns a list of `t` values that correspond to the self intersection points of the subpath. For each intersection point, the returned `t` value is the smaller of the two that correspond to the point.
+	/// - `error` - For intersections with non-linear beziers, `error` defines the threshold for bounding boxes to be considered an intersection point.
+	/// - `minimum_seperation`: the minimum difference two adjacent `t`-values must have when comparing adjacent `t`-values in sorted order.
+	/// If the comparison condition is not satisfied, the function takes the larger `t`-value of the two
+	///
+	/// **NOTE**: if an intersection were to occur within an `error` distance away from an anchor point, the algorithm will filter that intersection out.
+	// TODO: add iframe to rustdocs
```
Comment (RobNadal) [inline libraries/bezier-rs/src/subpath/solvers.rs:63]: You can add the iframe as part of this review, just follow the format of the others

File: website/other/bezier-rs-demos/src/features/subpath-features.ts [hunk 3]
```diff
@@ -40,32 +40,52 @@ const subpathFeatures = {
 		sliderOptions: [tSliderOptions],
 		chooseTVariant: true,
 	},
-	"intersect-linear": {
+	"Intersect (Line Segment)": {
```
Comment (RobNadal) [inline website/other/bezier-rs-demos/src/features/subpath-features.ts:43]: The shorthand name of the feature should not be modified (here, and for the other intersect features)

File: libraries/bezier-rs/src/subpath/solvers.rs (hunk 1 above)
Comment (Androxium) [reply libraries/bezier-rs/src/subpath/solvers.rs:20]: By removed, do you mean replace all the different match branches with 1 default match branch `_ => todo!(),`? The reason I changed this assignment to a match statement is because `intersection` and `self-intersection` return `(segment_idx: usize, t_value: f64)` while other subpath functions return `GlobalParametric(t_value)`.

Comment (RobNadal) [reply libraries/bezier-rs/src/subpath/solvers.rs:20]: I don't see why the match statement is necessary. `self.t_value_to_parametric(t)` will do the conversion from the different variants to a `(segment, index)` tuple, so there is no need to replicate within `evaluate`

File: libraries/bezier-rs/src/subpath/solvers.rs [hunk 4]
```diff
@@ -16,29 +16,24 @@ impl Subpath {
...
-		intersection_t_values.iter().fold(Vec::new(), |mut accumulator, t| {
-			if !accumulator.is_empty() && (accumulator.last().unwrap() - t).abs() < minimum_seperation.unwrap_or(MIN_SEPERATION_VALUE) {
+		intersection_t_values.iter().fold(Vec::<(usize, f64)>::new(), |mut accumulator, (index, t)| {
+			if !accumulator.is_empty() && (accumulator.last().unwrap().1 - t).abs() < minimum_seperation.unwrap_or(MIN_SEPERATION_VALUE) {
 				accumulator.pop();
 			}
-			accumulator.push(*t);
+			accumulator.push((*index, *t));
 			accumulator
 		});
```
Comment (RobNadal) [inline libraries/bezier-rs/src/subpath/solvers.rs:38]: Thinking about this a bit more, I'm not sure that the fold here is necessary.
1. Doesn't fold return a value? We're not actually capturing the return value of the fold, so I'm not confident that this line is actually doing anything. I don't think fold actually modifies things in place
2. The fold behavior is basically removing points that are within `MIN_SEPERATION_VALUE` of each other. But, doesn't Bezier::intersection already do that? The only points that wouldn't be `MIN_SEPERATION_VALUE` sequential are those that might be on different segments, but we don't really care about that case, the whole point of adding `MIN_SEPERATION_VALUE` is to neatly handle the case where two segments overlap and generate too many intersections.

TLDR: I think we can just remove the whole `intersection_t_values.iter().fold(...);`

**Review by Keavon — APPROVED** [review]
LGTM, just these couple comments.

File: libraries/bezier-rs/src/subpath/solvers.rs [hunk 5]
```diff
@@ -52,7 +39,32 @@ impl Subpath {
 		self.get_segment(segment_index).unwrap().tangent(TValue::Parametric(t))
 	}
 
-	/// Returns a normalized unit vector representing the direction of the normal on the subpath based on the parametric `t`-value provided.
```
[the remark this answers is not in the document]
Comment (Androxium) [reply libraries/bezier-rs/src/subpath/solvers.rs:55]: Oops, that's my bad, must have accidentally deleted it when rebasing.
